### ISSUE TITLE
Performance optimizations in X-MCFX

### ISSUE THREAD

**Issue by abhayap** [kronihias/mcfx#11]
Hi Matthias, I was wondering if it's possible to bring in the performance optimizations in [X-MCFX](http://www.angelofarina.it/X-MCFX.htm) if they are applicable?

**Comment by kronihias** [comment]
x-mcfx is forked off from mcfx, i didn't do any performance optimizations to mcfx after the fork happened (apart from adjusting for Apple Silicon)

according to the publication, x-mcfx supposedly has performance improvements, however, i wasn't able to reproduce this.

anyway, you can do the same things with mcfx as with x-mcfx.

**Comment by Angelo-Farina** [comment]
Hi Matthias, as you probably are aware, my research fellow Domenico Stefani worked on a fork of MCFX Convolver which makes it easier to deal with FIR filter matrices., He added to the GUI  a master gain, and the capability to inject metadata in  the WAV file containing the filter matrix, informing MCFX Convolver of the number of input channels for a full matrix, and a flag for telling when it is a diagonal matrix. Here the link to Domenico's fork: 
https://github.com/domenicostefani/mcfx-mod
It would be nice if you consolidate this as the official release 0.6.4 of MCFX, as these improvements are very useful. 
Domenico did not touch the convolution engine, as also in his tests the variant X-MCFX (https://github.com/JB-Luke/X-MCFX) developed by my previous research assistant, Luca Battisti (JB-Luke), did not provide significant speed advantages. 
In realty, after the latest fork of MCFX convolver was released by Domenico, I did a number of performance tests on various machines (not on Apple Silicon, as we never managed to compile X-MCFX for it.
On true last-gen Intel processors, with massively large FIR filter matrices, such as the beamforming Ambix 5th order filter matrix for the Eigenmike-64, which has 64 inputs and 36 outputs, the X-MCFX version is twice faster than MCFX. 
However, with smaller FIR filter matrices, and on AMD processors, such a large improvement does not appear. So I concluded that the X-MFX Convolver modified convolution engine is really faster than your original one, but only for certain filter matrices and on certain processor.
Now, I do not know if these speed improvements are worth the BIG effort required for realigning the code in the standard release of MCFX Convolver with the code in X-MCFX. 
Domenico evaluated that it was NOT worth the effort, and so possibly the same conclusion applies also to you.
So please, just bring online for everyone the Domenico Stefani forked version, with improved GUI and no modification to the critical convolution engine. During the next year, in 2025, I probably will have the resources for having another research assistant (or possibly Domenico again) working on a  further improved version of MCFX, re-employing the iX-MCFX advanced convolution engine, whilst retaining full compatibility with your master version.
But I cannot promise this just now, so, please, for now, just consolidate current Domenico's version, for the benefit of all your users.

**Comment by kronihias** [comment]
Hi @Angelo-Farina , thanks for the note.
I did incorporate the changes from https://github.com/domenicostefani/mcfx-mod.

Next time it would be great to have the authors make pull requests, open issues/tickets, and we can have discussions on the code changes and propose adjustments. To avoid problems such as the fork going too far in a direction that is not in-line with the ideas of the maintainer.

**Comment by Angelo-Farina** [comment]
Thanks Mathias!
I agree entirely on your suggestion, in fact it is exactly what I did say to Domenico at the beginning of his work..
And I did understand that Domenico had some interaction with you, but perhaps I misunderstood, as Domenico had a double task: one was fixing the GUI of MCFX, but his real job was to transplant the MCFX convolution engine (which I think that was originally coming from Fons' J-conv) inside the Sparta_6DOF_Convolver plugin, which we needed to use for the auralization of the Tyndaris theatre in the PHE project.
And here below is the result of this work, which works perfectly, and with smaller CPU load and faster and smoother position switching than the original, buggy software released by the Aalto University.
So perhaps I did misunderstand, and Domenico interacted with Leo Mc-cormack at Aalto...
Who, I hope, will also incorporate Domenico's work in the next release of Sparta plugins.
https://github.com/domenicostefani/SPARTA_Multi6DoF_project
Thanks, bye!

Angelo Farina
________________________________

**Comment by domenicostefani** [comment]
Hi Matthias and Angelo,
Great! I wanted to go for the PR but had been super busy and never got around to doing it.
Yes, the mods were minimal, and I mainly did those to get comfortable with the code before the engine-swap with SPARTA.

- For MCFX it seems you already merged what was needed right? Otherwise, I can go ahead with a PR.
- For SPARTA: The way I integrated MCFXconv was result-first focused and I had to forego reasoning on the cleanest way to integrate everything. However, now that it works I would like to pick your brain on how to refine the integration if you think it is needed. I would have loved to include MCFXConv as a GitHub module, but I had to change a few lines of code to allow swapping preallocated filters without allocation. It sounds like I would need to fork, apply the mods, and import that module, which does not sound amazing. Maybe, as the mods are minimal and not breaking, we could merge this with the main MCFX repo and use that.
Plus, I don't know the practical details of how to import the full MCFX repo, but just include&compile the few files relative to the Conv along with SPARTAs 6DoFconv.
I don't think there are intentions from SPARTA's people to integrate this in their releases. Still, it could help future developments as it would keep the two repositories tied. 
( e.g., It would be cool and not terribly hard to add multisource support to Sparta's plugin now)

**Comment by Angelo-Farina** [comment]
Exactly. In the next year, our  plan is to make Domenico's version of Sparta 6DOF convolver capable of full matrix convolution.
This will allow for directional Ambisonics sources., as the Iem RoomEncoder plugin..
Bye
Angelo

Angelo Farina
________________________________

**Comment by kronihias** [comment]
I did some more benchmarking and found those changes do reduce CPU load.

They are integrated in the latest release:

https://github.com/kronihias/mcfx/releases/tag/v0.7.1

> Dedicated to Angelo Farina, whose pioneering work on multichannel audio, Ambisonics, and room acoustics measurement was a great inspiration for these plug-ins.
